.github/workflows/test.yml

```diff
@@ -69,119 +65,108 @@ jobs:
-            --platform ${DOCKER_PLATFORMS// /,} \
-            --push \
-            -f cmd/shfmt/Dockerfile \
-            "$@" \
-            .
-        }
-
-        buildx -t "$DOCKER_BASE:$TAG"
-        buildx -t "$DOCKER_BASE:$TAG-alpine" --target alpine
-    - name: Test multi-architecture Docker images
-      run: |
-        printf '%s\n' "#!/bin/sh" "echo 'hello world'" >myscript
-
-        for platform in $DOCKER_PLATFORMS; do
-          for ext in '' '-alpine'; do
-            image="${DOCKER_BASE}:${TAG}${ext}"
-            msg="Testing docker image $image on platform $platform"
-            line="${msg//?/=}"
-            printf "\n${line}\n${msg}\n${line}\n"
-            docker pull -q --platform "$platform" "$image"
-            if [ -n "$ext" ]; then
-              echo -n "Image architecture: "
-              docker run --rm --entrypoint /bin/sh "$image" -c 'uname -m'
-            fi
-            version=$(docker run --rm "$image" --version)
-            echo "shfmt version: $version"
-            if [ "$TAG" != 'latest' -a "$TAG" != "$version" ]; then
-              echo "Version mismatch: shfmt $version tagged as $TAG"
-              exit 1
-            fi
-            docker run --rm -v "$PWD:/mnt" -w '/mnt' "$image" -d myscript
+      - name: Checkout code
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0 # also fetch tags for 'git describe'
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_USER }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+      - name: Set up image tag
+        run: |
+          set -vx
+          # Export environment variable for later stages.
+          if echo "$GITHUB_REF" | grep -q '^refs/heads/'; then
+            # Pushes to (master) branch - deploy 'latest'.
+            echo "TAG=latest" >> $GITHUB_ENV
+          elif echo "$GITHUB_REF" | grep -q '^refs/tags/'; then
+            # Pushes tag - deploy tag name.
+            echo "TAG=${GITHUB_REF/refs\/tags\//}" >> $GITHUB_ENV
+          fi
+          echo "DOCKER_BASE=${{ secrets.DOCKER_USER }}/${{ env.DOCKER_REPO }}" >> $GITHUB_ENV
+      - name: Build and push
+        id: buildx
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: ./cmd/shfmt/Dockerfile
+          platforms: ${{ env.DOCKER_PLATFORMS }}
+          push: true
+          tags: ${{ env.DOCKER_BASE }}:${{ env.TAG }}
+      - name: Build and push (alpine)
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: ./cmd/shfmt/Dockerfile
+          platforms: ${{ env.DOCKER_PLATFORMS }}
+          push: true
+          tags: ${{ env.DOCKER_BASE }}:${{ env.TAG }}-alpine
+          target: alpine
+      - name: Test multi-architecture Docker images
```

Comment:
Ideally this would also be reorganized. What the current flow does is:
* build and push the image to dockerhub
* test the image from dockerhub to check if it works
* if there is a problem, an error is reported, however, the bogus image is already uploaded to dockerhub

It seems that there is built-in support now that makes it easy to [spin up a local temporary docker registry for testing purposes](https://github.com/docker/build-push-action#local-registry). So the better approach would be:

* build and push the image to a local temporary docker registry
* test the image from the local docker registry to check if it works
* if there is a problem report an error and quit
* if the tests were successful, push the image to dockerhub

Reply:
Again, yes! Fully agree. Let's try to implement this in a next PR. Also, I would suggest splitting the workflows in multiple files to facilitate maintenance.